### PR TITLE
Vectorize (CPU) generic types for binary bitwise operators

### PR DESCRIPTION
**Pull request by xuhdev**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#34338 Vectorize (CPU) generic types for binary bitwise operators**

For those types not optimized for AVX2, this commit would give bitwise
operations on them a boost.

Benchmark (RHEL 7.7, Intel(R) Core(TM) i7-8850H CPU @ 2.60GHz, Turbo off, Release build):

```python
import timeit
for op in ('bitwise_and', 'bitwise_or', 'bitwise_xor'):
    for dtype in ('torch.int8', 'torch.uint8'):
        for n, t in [(10_000, 200000),
                    (100_000, 20000)]:
            print(f'a.{op}_(b), numel() == {n} for {t} times, dtype={dtype}')
            print(timeit.timeit(f'a.{op}_(b)', setup=f'import torch; a = torch.arange(1, {n}, dtype={dtype}); b = torch.arange({n}, 1, -1, dtype={dtype})', number=t))
```

Before:

```
a.bitwise_and_(b), numel() == 10000 for 200000 times, dtype=torch.int8
1.353799690001324
a.bitwise_and_(b), numel() == 100000 for 20000 times, dtype=torch.int8
1.056434961999912
a.bitwise_and_(b), numel() == 10000 for 200000 times, dtype=torch.uint8
1.2957618809996347
a.bitwise_and_(b), numel() == 100000 for 20000 times, dtype=torch.uint8
1.0591609650000464
a.bitwise_or_(b), numel() == 10000 for 200000 times, dtype=torch.int8
1.3113185389993305
a.bitwise_or_(b), numel() == 100000 for 20000 times, dtype=torch.int8
1.0693870880022587
a.bitwise_or_(b), numel() == 10000 for 200000 times, dtype=torch.uint8
1.3075691039994126
a.bitwise_or_(b), numel() == 100000 for 20000 times, dtype=torch.uint8
1.0589785859992844
a.bitwise_xor_(b), numel() == 10000 for 200000 times, dtype=torch.int8
1.3036618039986934
a.bitwise_xor_(b), numel() == 100000 for 20000 times, dtype=torch.int8
1.0595013140009542
a.bitwise_xor_(b), numel() == 10000 for 200000 times, dtype=torch.uint8
1.2947387999993225
a.bitwise_xor_(b), numel() == 100000 for 20000 times, dtype=torch.uint8
1.059969027999614
```

After:

```
a.bitwise_and_(b), numel() == 10000 for 200000 times, dtype=torch.int8
0.9562859639991075
a.bitwise_and_(b), numel() == 100000 for 20000 times, dtype=torch.int8
0.6811799210008758
a.bitwise_and_(b), numel() == 10000 for 200000 times, dtype=torch.uint8
0.9522694869992847
a.bitwise_and_(b), numel() == 100000 for 20000 times, dtype=torch.uint8
0.6815469840003061
a.bitwise_or_(b), numel() == 10000 for 200000 times, dtype=torch.int8
0.8609786279994296
a.bitwise_or_(b), numel() == 100000 for 20000 times, dtype=torch.int8
0.5794818879985542
a.bitwise_or_(b), numel() == 10000 for 200000 times, dtype=torch.uint8
0.8534434389985108
a.bitwise_or_(b), numel() == 100000 for 20000 times, dtype=torch.uint8
0.5764101290005783
a.bitwise_xor_(b), numel() == 10000 for 200000 times, dtype=torch.int8
0.9634105910008657
a.bitwise_xor_(b), numel() == 100000 for 20000 times, dtype=torch.int8
0.6819724230008433
a.bitwise_xor_(b), numel() == 10000 for 200000 times, dtype=torch.uint8
1.0901075929978106
a.bitwise_xor_(b), numel() == 100000 for 20000 times, dtype=torch.uint8
0.816546294001455
```

Differential Revision: [D20687081](https://our.internmc.facebook.com/intern/diff/D20687081)

### CODE REVIEW

File: aten/src/ATen/cpu/vec256/vec256_base.h [hunk 1]
```diff
@@ -606,33 +609,65 @@ Vec256<T> inline clamp_min(const Vec256<T> &a, const Vec256<T> &min_vec) {
   return c;
 }
 
-template <class T, typename Op, typename std::enable_if_t<!std::is_integral<T>::value, int> = 0>
+struct Vec256i;
+
+#ifdef __AVX2__
+
+template <class T, typename Op>
 static inline Vec256<T> bitwise_binary_op(const Vec256<T> &a, const Vec256<T> &b, Op op) {
```
Comment (ezyang) [inline aten/src/ATen/cpu/vec256/vec256_base.h]: This only works for integral types, right? I am confused why this modification wasn't put in `vec256_int.h`. Do you know why?

**Review by ezyang — CHANGES_REQUESTED** [review]
code structure problem

File: aten/src/ATen/cpu/vec256/vec256_base.h (hunk 1 above)
Comment (xuhdev) [reply aten/src/ATen/cpu/vec256/vec256_base.h]: I think it should work for any types that are not AVX2-optimized? All it does it to take advantage of int-type avx2 instructions.

File: aten/src/ATen/cpu/vec256/vec256_base.h [hunk 2]
```diff
@@ -606,33 +609,65 @@ Vec256<T> inline clamp_min(const Vec256<T> &a, const Vec256<T> &min_vec) {
...
-  using iT = int_same_size_t<T>;
-  iT buffer[Vec256<T>::size()];
-  for (int i = 0; i != Vec256<T>::size(); i++) {
-    auto a_val = a[i];
-    auto b_val = b[i];
-    iT *i_a_ptr = reinterpret_cast<iT*>(&a_val);
-    iT *i_b_ptr = reinterpret_cast<iT*>(&b_val);
+  __m256i buffer;
+  __m256i a_buffer = _mm256_loadu_si256(reinterpret_cast<const __m256i*>((const T*)a));
```
[the remark this answers is not in the document]
Comment (xuhdev) [reply aten/src/ATen/cpu/vec256/vec256_base.h]: I guess you might have misunderstood. This change is to improve the base class, not only for integers. The optimization in your mind might have been done in this template line:

https://github.com/pytorch/pytorch/pull/34338/files#diff-820afb643a1ef3fe10d40fde5c919060R702

File: aten/src/ATen/cpu/vec256/vec256_base.h (hunk 1 above)
Comment (ezyang) [reply aten/src/ATen/cpu/vec256/vec256_base.h]: Err, so I agree you can technically bitwise OR a float, by treating it as some bits and ORing the bits. But I find it hard to believe that anyone would want that. Is that what you mean?

Comment (xuhdev) [reply aten/src/ATen/cpu/vec256/vec256_base.h]: That's what I meant. The bitwise operators actually are already available for floating-point numbers on AVX2-enabled CPUs: https://github.com/pytorch/pytorch/blob/2d24005d186844ace9020aa94a5be293f6396f16/aten/src/ATen/cpu/vec256/vec256_float.h#L304-L317

I simply don't see any reason to limit this optimization to non-optimized integer vectors, given the fact that there isn't much difference in the code. There may be potentially creative use of other non-optimized types, such as `half` and `bfloat` (for example changing the sign bit).

Comment (ezyang) [reply aten/src/ATen/cpu/vec256/vec256_base.h]: OK. In that case if you can show me evidence that we aren't doing unnecessary extra loads in the case of integer vectors (e.g., show me the assembly after this change) I'll accept it in this style.

Comment (xuhdev) [reply aten/src/ATen/cpu/vec256/vec256_base.h]: OK. Would this answer your question: https://godbolt.org/z/PTdzXU The new implementation seems to lead to much more optimized assemblies.

Comment (ezyang) [reply aten/src/ATen/cpu/vec256/vec256_base.h]: So, first, I agree that the new code is better than what we did before.

But I think it could be even better. Here's why.

We write operations directly on `Vec256` because that makes things compositional. So most of the operations in this class don't do explicit loads/stores to the AVX registers, because we assume that things are already in `__m256`. But your code here violates this assumption (ostensibly, because you are doing the cast), *despite* the fact that you are being passed in `Vec256`.

This seems to matter in practice. Here is a modified version of your Godbolt: https://godbolt.org/z/hycFeD new_bitwise_and3 shows an implementation of bitwise and between three operands. I've shown two versions: one where I implement it using your code (which does the load/store), and another "optimal" version where I write everything by hand. The extra loads and stores show up in the generated assembly.

Here is my claim: if you had put this implementation in the integer VEc256 specialization, this would have removed the necessity for doing the reinterpret cast (which is probably why you were doing the load/store in the first place). Then you could have operated directly on the passed in `__m256i` registers. Then your implementations would have been compositional in the way I describe.

Comment (xuhdev) [reply aten/src/ATen/cpu/vec256/vec256_base.h]: It seems to me that we still have some misunderstanding.

1. My main claim is that all integer Vec256 specializations won't be affected by this PR.

> if you had put this implementation in the integer VEc256 specialization, this would have removed the necessity for doing the reinterpret cast (which is probably why you were doing the load/store in the first place).

All bitwise operations on Vec256 that inherits Vec256i (which uses `__m256i` without cast) have been implemented already:

https://github.com/pytorch/pytorch/blob/12fb8148e48d3e39d16832a84ac4b59182a00c4e/aten/src/ATen/cpu/vec256/vec256_int.h#L694-L705

This PR does not delete these implementations and doesn't override any of these. The added functions explicitly excluded situations that Vec256 inherits Vec256i (see the template declaration above the newly defined bitwise operator overloads):

    template<class T, typename std::enable_if_t<!std::is_base_of<Vec256i, Vec256<T>>::value, int> = 0>

Therefore, in case AVX2 is enabled, this PR should only have changed the behavior when there is no specifically optimized AVX2 specialization of `Vec256<T>`, such as `Vec256<int8_t>`, `Vec256<uint8>`, `Vec256<bfloat16>`, etc., until a specifically optimized specialization is implemented.

2. My secondary claim is that the load/store operations won't appear in the assemblies that this PR would generate.

I made a mistake in the previous godbolt link. If the arrays are aligned (as what this PR does), the load and store instructions are gone in the assembly:

https://godbolt.org/z/Jbg7Ke

(See the difference between `new_bitwise_and_noalign` and `new_bitwise_and`.)

Comment (ezyang) [reply aten/src/ATen/cpu/vec256/vec256_base.h]: Ahhh, here we go! OK, this is good enough for me.

**Review by ezyang — APPROVED** [review]

**Comment by xuhdev** [comment]
Any chance to get this merged :) It's been quite a while
